staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go

```diff
@@ -0,0 +1,454 @@
+		kubectl alpha debug --image=myproj/debug-tools -c debugger mypod`))
+
+	errDisabled = "ephemeral containers are disabled for this cluster (\"%s\")."
+)
+
+var nameSuffixFunc = utilrand.String
+
+// Options holds the options for an invocation of kubectl debug.
+type Options struct {
+	Args        []string
+	Attach      bool
+	Command     bool
+	Container   string
+	Env         []corev1.EnvVar
+	Image       string
+	Interactive bool
+	Namespace   string
+	PodNames    []string
+	PullPolicy  corev1.PullPolicy
+	Quiet       bool
+	Target      string
+	TTY         bool
+
+	builder   *resource.Builder
+	podClient corev1client.PodsGetter
+
+	genericclioptions.IOStreams
+}
+
+// NewOptions returns a Options initialized with default values.
+func NewOptions(streams genericclioptions.IOStreams) *Options {
+	return &Options{
+		IOStreams: streams,
+		Args:      []string{},
+		PodNames:  []string{},
+	}
+}
+
+// NewCmdDebug returns a cobra command that runs kubectl debug.
+func NewCmdDebug(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
+	o := NewOptions(streams)
+
+	cmd := &cobra.Command{
+		Use:                   "debug NAME --image=image -- [COMMAND] [args...]",
+		DisableFlagsInUseLine: true,
+		Short:                 i18n.T("Attach a debug container to a running pod"),
+		Long:                  debugLong,
+		Example:               debugExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(o.Complete(f, cmd, args))
+			cmdutil.CheckErr(o.Validate(cmd))
+			cmdutil.CheckErr(o.Run(f, cmd))
+		},
+	}
+
+	addDebugFlags(cmd, o)
+	return cmd
+}
+
+func addDebugFlags(cmd *cobra.Command, opt *Options) {
+	cmd.Flags().StringVar(&opt.Container, "container", opt.Container, i18n.T("Container name to use for debug container."))
+	cmd.Flags().StringVar(&opt.Image, "image", opt.Image, i18n.T("Container image to use for debug container."))
+	cmd.MarkFlagRequired("image")
+	cmd.Flags().String("image-pull-policy", string(corev1.PullIfNotPresent), i18n.T("The image pull policy for the container."))
+	cmd.Flags().StringToString("env", nil, "Environment variables to set in the container.")
+	cmd.Flags().StringP("labels", "l", "", "Comma separated labels to apply to the pod(s). Will override previous values.")
+	cmd.Flags().BoolVarP(&opt.Interactive, "stdin", "i", opt.Interactive, "Keep stdin open on the container(s) in the pod, even if nothing is attached.")
+	cmd.Flags().BoolVarP(&opt.TTY, "tty", "t", opt.TTY, "Allocated a TTY for each container in the pod.")
+	cmd.Flags().BoolVar(&opt.Attach, "attach", opt.Attach, "If true, wait for the Pod to start running, and then attach to the Pod as if 'kubectl attach ...' were called.  Default false, unless '-i/--stdin' is set, in which case the default is true.")
+	cmd.Flags().BoolVar(&opt.Command, "command", opt.Command, "If true and extra arguments are present, use them as the 'command' field in the container, rather than the 'args' field which is the default.")
```

Comment:
This is confusing, I'd suggest doing it other way, by default everything goes into `Command` which is the default with `kubectl exec` and `kubectl attach` and if you want to pass only arguments `--arguments-only` and consume everything after `--` as arguments only.

Reply:
I just copied this behavior from `kubectl run`, but I agree it's confusing. `kubectl exec` is a little different because there's no default command specified in the image like there is here. I find `--arguments-only` to be a bit confusing, too.

When looking into this, I came across the [kubernetes.io docs for command vs args](https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/), and I think we should try to stick to the spirit of these here. As with docker, there's a slight bias for args over entrypoint here.

I've rewritten this to simply allow specifying a command, but otherwise defaulting to only using args. My goal is to make it work in a similar fashion to `docker run`:

```
kubectl debug mypod -it --image=debian
kubectl debug mypod -it --image=debian -- sh
kubectl debug mypod -it --image=debian -- echo foo bar
kubectl debug mypod -i --image=debian --command=/bin/sh -- -c 'echo foo $HOME'
kubectl debug mypod --image=nmap -- -sT localhost
```

Reply:
In all honesty, I don't care much about being similar to `docker run`, and as far as it does make sense for `kubectl run`, it's different with `kubectl debug` imho. The only reason I need to jump and debug a problem is when something is not working as expected, passing arguments will be the last thing I'll be interested, but being able to specify a specific command is more. I bet users will be confused and one of the first issues will be to change that to what I'm saying here. Other `kubectl` commands follow that approach and I'd prefer we keep it consistent. I'm not tirelessly defend `--arguments-only` since that's just naming scheme, can be different, we can inject arguments with that flag, but I'd strongly prefer this approach over the current one.

Reply:
As discussed in slack, I'm happy to defer to your experience and judgement here. I've implemented your original `--arguments-only` suggestion. As we add functionality to debug maybe we'll revisit whether we even need to support Args at all, or whether only supporting Command would suffice.